drivers/bus/acpi/eval.c

```diff
@@ -1,255 +1,804 @@
+#ifndef UNIT_TEST
 #include "precomp.h"
 
 #define NDEBUG
 #include <debug.h>
+#endif /* UNIT_TEST */
+
+#define AcpiVerifyInBuffer(Stack, Length) \
+    ((Stack)->Parameters.DeviceIoControl.InputBufferLength >= Length)
+
+#define AcpiVerifyOutBuffer(Stack, Length) \
+    ((Stack)->Parameters.DeviceIoControl.OutputBufferLength >= Length)
+
+#define TAG_ACPI_PARAMETERS_LIST     'OpcA'
+#define TAG_ACPI_PACKAGE_LIST        'PpcA'
+#define TAG_ACPI_OUTPUT_BUFFER       'BpcA'
+
+/* Null terminated ACPI name for the object */
+#define ACPI_OBJECT_NAME_LENGTH      (4 + 1)
+
+/* Maximum number of nested package structures supported by the driver */
+#define ACPI_MAX_PACKAGE_DEPTH       5
+
+static
+CODE_SEG("PAGE")
+NTSTATUS
+EvalConvertObjectReference(
+    _Out_ PACPI_METHOD_ARGUMENT Argument,
+    _In_ ACPI_OBJECT* Reference)
+{
+    ACPI_BUFFER OutName;
+    ACPI_STATUS AcpiStatus;
+
+    PAGED_CODE();
+
+    Argument->Type = ACPI_METHOD_ARGUMENT_STRING;
+    Argument->DataLength = ACPI_OBJECT_NAME_LENGTH;
+
+    /* Convert the object handle to an ACPI name */
+    OutName.Length = ACPI_OBJECT_NAME_LENGTH;
+    OutName.Pointer = &Argument->Data[0];
+
+    AcpiStatus = AcpiGetName(Reference->Reference.Handle, ACPI_SINGLE_NAME, &OutName);
+    if (!ACPI_SUCCESS(AcpiStatus))
+    {
+        DPRINT1("AcpiGetName() failed on %p with status 0x%04lx\n",
+                Reference->Reference.Handle,
+                AcpiStatus);
+
+        ASSERT(FALSE);
+        return STATUS_UNSUCCESSFUL;
+    }
+
+    return STATUS_SUCCESS;
+}
 
 static
+CODE_SEG("PAGE")
 NTSTATUS
-GetPackageSize(ACPI_OBJECT *Package,
-               PULONG Count,
-               PULONG Size)
+EvalGetElementSize(
+    _In_ ACPI_OBJECT* Obj,
+    _In_ ULONG Depth,
+    _Out_opt_ PULONG Count,
+    _Out_ PULONG Size)
 {
-    ULONG Length, RawLength, TotalLength;
+    ULONG TotalCount, TotalLength;
+    NTSTATUS Status;
     UINT32 i;
```

Comment:
`i` could be moved into the `case ACPI_TYPE_PACKAGE:` block.

Reply:
Done.